Enhancing Dependency Injection and Reconnection logic

kum-deepak opened the pull request:
Reconnecting:

- Switched to @stomp/stompjs, which provides automatic reconnects.
- Since, I maintain that version, I was able to add the type definitions to that NPM package, so no need to add the type definitions separately.

Dependency Injection:

- The configuration is now injected directly into StompService instead of being set using a method call by another component - better loose coupling.
- In real-life there may be multiple components which may want to subscribe to different queues for different purposes. To support this, subscription queue name is no longer part of initial configuration, it is passed as call to subscribe. This method now returns an observable of messages for this queue.
- A queue can be subscribed even before Stomp connection has been established. Other components need not check if Stomp is connected.
- In case of reconnects the same observable will continue serving messages for the queue - will make life of queue subscribers much easier.
- Publishing takes the queue name as argument, to allow different components to publish to different queues.

What I am planning to work on next:

- Currently StompService has hard coded URL for configuration, need to convert to Injectable Dependency.
- Publish should check if it is currently connected, if not it should locally store and send when connected.
- Quite often manual ack is needed for incoming messages, need to expose that to queue subscribers.
- Convert it to Angualr2 NPM package.

kum-deepak committed:
Working with RabbitMQ and @stomp/stompjs

kum-deepak committed:
Reconnect logic from @stomp/stompjs

kum-deepak committed:
Refactor 01: Moved Config Service to where it should be.

kum-deepak committed:
Refactor 02: Better Inversion of Control

kum-deepak committed:
Refactor 03: Reconnect delay in config

kum-deepak committed:
Refactor 04: Convert to use Observer instead of Subject.

sjmf commented:
Hi @kum-deepak - thanks for this PR.

Your changes look good, but the PR is based on the `reconnect` branch which is 4 commits behind master. Could you please merge in the changes from master to resolve conflicts?

kum-deepak committed:
Merge branch 'master' into reconnect

# Conflicts:
#	src/app/services/stomp/stomp.service.ts

kum-deepak committed:
Merged all changes from current master branch of https://github.com/sjmf/ng2-stompjs-demo.git

kum-deepak commented:
Merged. The current config works with RabbitMQ.

JimiC commented:
I have some comments but will not be able to fully review it today. Gimme some time to do it during the weekend.

kum-deepak committed:
Implemented concept of unsubscribe, I think I have gotten Angular2 like API.

kum-deepak commented:
Added concept of Unsubscribe. Stop the Observable subscriptions and it unsubscribes from Stomp Broker. There was also a subtle bug, it was creating two subscriptions at Stomp broker if the Observable was subscribed twice (as was the case in the current example).

I have updated the type definitions in the stompjs code - the subscribe function in stompjs actually returns an object which has unsubscribe method.

kum-deepak committed:
Implemented local queueing of messages if Stomp borker is not connected. These will
be sent when the borker connects subsequently.

kum-deepak commented:
I have now implemented local queuing of messages if Stomp broker is not connected. 

This leaves only one more change that I had planned - removal of hard coded URL from the config system. That is a relatively small change, so, will do that when other changes are final.

sjmf commented on src/app/services/stomp/stomp.service.ts:
```diff
@@ -139,28 +128,89 @@ export class STOMPService {
 
-    for (const t of this.config.publish) {
-      this.client.send(t, {}, message);
+  /** Send a message, locally quque if not connected */
```
"locally queue"

kum-deepak replied:
Thanks :)

sjmf commented on src/api/config.json:
```diff
@@ -1,16 +1,16 @@
 {
```
I'd appreciate your thoughts on having a default config which 'just works'. I'd previously written it so that devs cloning the repo would have to update the `config.json` first, to point it to their MQ instance (hence `example.com.invalid`, which is non-routable). Your proposed `config.json` will connect to a RabbitMQ instance with the default user running on localhost, which seems like a better idea 👍 - can we think of any downsides? (I'm trying to remember why I did it that way in the first place)

kum-deepak replied:
We can actually provide more than one sample config.json, corresponding to few popular brokers. My simple reasoning - many people like instant gratification :)

sjmf commented on src/app/components/rawdata/rawdata.component.ts:
```diff
@@ -25,44 +29,57 @@ export class RawDataComponent implements OnInit, OnDestroy {
       return Math.floor(Math.random() * (max - min + 1)) + min;
     };
-    this._stompService.publish(`{ type: "Test Message", data: [ ${this._counter}, ${_getRandomInt(1, 100)}, ${_getRandomInt(1, 100)}] }`);
+    this._stompService.publish("/topic/ng-demo-sub", `{ type: "Test Message", data: [ ${this._counter}, ${_getRandomInt(1, 100)}, ${_getRandomInt(1, 100)}] }`);
```
Hard-coded topic for demo: Is there a rationale for this, or should this be parameterized (and/or in `config.json`?)

EDIT: Ah, I just re-read your comment above. Shall we make this change part of this PR? Thanks.

kum-deepak replied:
It should definitely not be in config.json. In actual applications, it will either be the application configuration or will be hard coded in the component.

Currently this sample shows both the service code and the sample code that uses it. Any of the queue names are definitely not part of the service.

In my opinion, for this sample it can be hard coded.

sjmf commented on src/app/services/stomp/stomp.service.ts:
```diff
@@ -23,53 +24,55 @@ export enum STOMPState {
+
+    // Setup sending queuedMessages
+    this.connectObservable.subscribe(() => {
+      // The delay is just for testing during development, as I am publishing and subscribing same Stomp topic. the delaye
```
Delay for testing during development: was it your intention to include this as part of the PR? 

If my understanding is correct, it sounds like perhaps this should either be a permanent feature, or the delay be removed?

kum-deepak replied:
Thanks, removed it. Only one caution, in current demo, I had configured same queue names for publish and subscribe. On reconnect, in all  likelihood, on a reconnect, messages will be published before the subscription gets reestablished.

In actual production code, this will pose no issue.

Though we will need to make it explicit in our Readme, so that people do not report it as a bug.

sjmf commented on src/app/services/stomp/stomp.service.ts:
```diff
@@ -139,28 +128,89 @@ export class STOMPService {
 
-  /** Subscribe to server message queues */
-  public subscribe(): void {
+    this.debug(`Will try sending ququed messages ${queuedMessages}`);
```
"ququed"?

kum-deepak replied:
Thanks :)

sjmf commented on src/app/services/stomp/stomp.service.ts:
```diff
@@ -139,28 +128,89 @@ export class STOMPService {
+     * to this observable twice, it will subscribe twice to Stomp broker. (This was happening in the current example).
+     * A long but good explanatory article at https://medium.com/@benlesh/hot-vs-cold-observables-f8094ed53339
+     */
+    return coldObservable.share();
```
I love this feature addition. <3

kum-deepak replied:
Really appreciated :)

sjmf reviewed (commented):
Thanks so much for this contribution. I've left a small number of comments which I'd like your feedback on before we get it merged in ^_^

Also, is the yarn lockfile (`yarn.lock`) necessary to include in the merge?

kum-deepak committed:
Updated based on review comments from https://github.com/sjmf

kum-deepak commented:
Regarding yarn.lock, it is not necessary to be included. For now, I will remove it from the commit.

It has some benefits though. It guarantees that anyone running the sample will use exactly same version of all the libraries if they also use yarn.

However, it is an independent decision to switch to using yarn.

Yarn has a nifty feature over 'NPM install', that it supports local file paths for libraries. This make life much easier when I have to update any library source (like in this case @stomp/stompjs).

kum-deepak committed:
Removed yarn.lock from commit.

JimiC commented:
**Disclaimer:** In general, I don't agree with some of the changes but if @sjmf is fine with them so am I.
From what I understood the purpose of this project is to provide an insight to the user (demo) on how to integrate a basic STOMP client with Angular and not provide an out-of-the-box complete functionality client.
Any excess functionality the user may want is up to the user to tweak the code and bring it to their liking.
From my POV it should be kept as simple, tweakable and as generic as possible.

*Config:*
The only addition I welcome is the `reconnect_delay`.
Every other property should stay as it was (except `https`, see below). 
Topics should not be hard-coded.
`path` property is redundant as RabbitMQ stomp protocol listens to `stomp/websocket` only. In my tests `ws` was not working.
I can't see the use case in having buttons to `subscribe/unsubscribe`. Is it to test your auto `subscribe/unsubscribe` functionality?

One bug I found and I'm to blame, is the use of `https` as a config property instead of `ssl`, as the code is looking for the latest to determine the `scheme`/`protocol` to use.

*Coding Guidelines:*
I noticed many cases where linting is not done.
You are using `let` where you should use `const`.

kum-deepak committed:
Almost lint clean.

kum-deepak commented:
Thanks Jimi!

_General:_
Yes, I do agree that it has taken a different approach, with a different purpose. This branch is to create a generically usable NPM module usable in any Angular2 (or higher) project without needing to tinker with the library code. In this effort the code has changed so much that it can very well be kept in a different branch :)

_Coding guidelines:_
The code is almost lint clean. I have not touched 3 cases on ;'s, which I wasn't sure about.

_Config:_
It is good idea to have path over here, as different Stomp brokers may use different path schemes. RabbitMQ also supports multiple paths (https://www.rabbitmq.com/web-stomp.html).

_Rawdata Component:_
The buttons are definitely not required. The concept of calling subscribe from init and unsubscribe from destroy need to be there. This will ensure the underlying queue subscription is released (Stomp unsubscribe) when the component is destroyed.

sjmf commented:
Looks like we're getting there. There's multiple changes in this PR which imo add a lot to the demo– in terms of the connection state machine– and actually it looks like a simplification in some terms.

I'm still sticking on the use of `'/topic/ng-demo-sub'` as a hard-coded topic.

sjmf commented:
Why was this closed? I just cloned it to do some final testing.

kum-deepak commented:
Sorry, my mistake, I was merging that to my local branch and by mistake deleted that.

Give me few minutes.

JimiC commented:
I was about to comment later in the day but what the heck here it is.
If the purpose of @kum-deepak was to make an `npm package` out of this repo then it deserved to be on its own repo and this repo could benefit from the proposed enhancements but stay as an educational one.

kum-deepak committed:
Almost lint clean.

kum-deepak commented:
Restored.

sjmf commented:
That's also a possibility, to create an `ng2-stompjs` package or similar, and back-port features. Thoughts? I'm open to suggestions.

kum-deepak commented:
That actually may be a good idea. I was planning to start work on that now.

JimiC commented:
Glad we came to an agreement.
@kum-deepak As the potential maintainer of `@stomp/stompjs` I suggest you host the `ng2-stompjs-package` on that account.

JimiC commented:
I just saw you made a repo there but I would suggest to name it `ng2-stompjs`.

kum-deepak commented:
Great, Just now created https://github.com/stomp-js/ng2-stomp

I will start committing in few days.

kum-deepak commented:
The NPM package name will be @stomp/ng2-stomp.

JimiC commented:
Because ppl would likely search by `ng2` and `stompjs`, `@stomp/ng2-stompjs` would be preferable.

kum-deepak commented:
Will change that, thanks!

kum-deepak commented:
Took liberty of adding both of you as collaborators. I hope it is ok.

pantonis commented:
@kum-deepak great work man. Stompjs is stable but abandoned so no support of TS or any other updates. Was looking for an angular 2 lib for stompjs. 
And I couldnt agree more with @JimiC and @sjmf that this repo should stay as it is. A demo for educational purposes. 

Will be waiting for your release.

JimiC commented:
@pantonis `Stompjs` has support for `TS` @ https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/stompjs

kum-deepak commented:
I have committed initial version at https://github.com/stomp-js/ng2-stompjs, this PR can now be closed :)

sjmf commented:
Thanks, closing :)
